channels/generic/multiplexer.py

```diff
@@ -0,0 +1,322 @@
+import asyncio
+import logging
+from functools import partial
+from typing import Any, Dict
+
+from channels.consumer import get_handler_name
+from channels.exceptions import StopConsumer
+from channels.generic.websocket import AsyncJsonWebsocketConsumer
+
+logger = logging.getLogger(__name__)
+
+
+class AsyncJsonWebsocketDemultiplexer(AsyncJsonWebsocketConsumer):
+
+    """
+    JSON-understanding WebSocket consumer subclass that handles demultiplexing
+    streams using a "stream" key in a top-level dict and the actual payload
+    in a sub-dict called "payload". This lets you run multiple streams over
+    a single WebSocket connection in a standardised way.
+    Incoming messages on streams are dispatched to consumers so you can
+    just tie in consumers the normal way.
+
+    Set a mapping of streams to an application classes in the "applications"
+    keyword.
+    """
+
+    # timeout in seconds after close used before killing child applications.
+    application_close_timeout = 10
+
+    # the mapping of applications by stream_name
+    applications = {}  # type: Dict[str: 'typing.Type'[channels.consumer.AsyncConsumer]]
+
+    def __init__(self, scope: Dict[str, Any]):
+        super().__init__(scope)
+        self.stream_upstream_queues = {}  # type: Dict[str, asyncio.Queue]
+        self.accepted_upstream_stream = set()
+        self.disconnecting = False
+        self.sent_close = False
+        self._task_monitors = None  # type: typing.Optional[asyncio.Task]
+
+    # Methods to INTERCEPT Client -> Stream Applications
+
+    async def websocket_connect(self, message):
+        """
+        Connect and then inform each stream application.
+        """
+        await super().websocket_connect(message)
+        for queue in self.stream_upstream_queues.values():
+            queue.put_nowait(message)
+
+    async def websocket_disconnect(self, message):
+        """
+        Handle when the connection is lost.
+        """
+
+        # set the disconnecting so that we know to to send frames further down.
+        self.disconnecting = True
+
+        for stream_name, queue in self.stream_upstream_queues.items():
+            # inform all upstream applications that the connection is going
+            #  down (this includes ones that never accepted!)
+            queue.put_nowait(message)
+
+        if self._task_monitors is None:
+            await super().websocket_disconnect(message)
+            return
+
+        try:
+            # wait for all the upstream applications to close down
+            await asyncio.wait_for(
```

Comment:
There's no need to wait for them to be killed. Just block forever and let the outside server kill your whole coroutine if things go bad.

Reply:
then we wont be able to raise any warnings at this level we will just get the warning from daphne?

Reply:
also will this not just result in there always being a warning being raised?